tests/foreman/longrun/test_puppet.py

```diff
@@ -0,0 +1,89 @@
+"""End to end test for Puppet funcionality
+
+@Requirement: Puppet
+
+@CaseAutomation: Automated
+
+@CaseLevel: Acceptance
+
+@CaseComponent: CLI
+
+@TestType: Functional
+
+@CaseImportance: High
+
+@Upstream: No
+"""
+
+from robottelo.config import settings
+from robottelo.decorators import (
+    run_in_one_thread,
+    run_only_on,
+    skip_if_not_set,
+    stubbed,
+    tier4,
+)
+from robottelo.test import CLITestCase
+
+
+@run_in_one_thread
+@skip_if_not_set('clients')
+class PuppetScenarioTestCase(CLITestCase):
+    """Implements Puppet test scenario"""
+
+    @classmethod
+    def setUpClass(cls):
+        """ Spin up Capsules and Clients.
+        """
+        super(PuppetScenarioTestCase, cls).setUpClass()
+        cls.sat6_hostname = settings.server.hostname
+        """
+        1. prepare for spinning up VMs (spinned up along testing)
+        """
+
+    @run_only_on('sat')
+    @stubbed
+    @tier4
+    def test_positive_puppet_scenario(self):
+        """
+        @id: fd311168-afda-49b6-ac5f-533c4fd411b5
+
+        @Steps:
+
+        1. register client to p3 satellite and apply motd module
+
+        2. prepare p3 capsule connected to p3 satellite
+
+        3. register client to p3 capsule and apply motd module
+
+        4. upgrade p3 capsule to p4
+
+        5. register client to upgraded p4 capsule and apply motd module
+
+        6. prepare p4 capsule connected to p3 satellite
+
+        7. register client to p4 capsule and apply motd module
+
+        8. upgrade p3 satellite to p4
+
+        9. register client to upgraded p4 satellite and apply motd module
+
+        10. prepare p3 capsule connected to upgraded p4 satellite
+
+        11. register client to p3 capsule and apply motd module
+
+        12. upgrade p3 capsule to p4
+
+        13. register client to upgraded p4 capsule and apply motd module
+
+        14. prepare p4 capsule connected to upgraded p4 satellite
+
+        15. register client to p4 capsule and apply motd module
+
```

Comment:
Just wondering about the puppet version on the registered clients, will we assume that the client will have the same P version as the capsule/satellite it is registered to? Also I guess we'll have to do some cleanups of the motd module between some steps.